### PR TITLE
[WIP] Complete rewrite of Windows exec wrapper

### PR DESCRIPTION
**Pull request by nitzmahone**
##### ISSUE TYPE
 - Feature Pull Request

##### SUMMARY
Complete rewrite of Windows exec wrapper    
* supports pipelining for faster execution
* supports become (runas), creates interactive subsession under WinRM batch logon
* supports usage of arbitrary module_utils files
* modular exec wrapper payload supports easier extension
* integrates async wrapper behavior for pipelined/become'd async
* module_utils are loaded as true Powershell modules, no more runtime modifications to module code

### CODE REVIEW

**Comment by jhawkesworth** [comment]
Awesome, looking forward to testing this out as soon as I can.   Many, many thanks

**Comment by dagwieers** [comment]
Whoo, much anticipated !

**Comment by jborean93** [comment]
Looking forward to doing some testing when it is ready, are you able to briefly explain what you are doing with the pipelining work and what that actually means with how Ansible executes scripts?

**Comment by dagwieers** [comment]
Seems to work fine in my tests.

**Comment by abadger** [comment]
+1 from me

**Comment by nitzmahone** [comment]
@jborean93 Sure: pipelining means the module never hits the disk. The classic Ansible module exec pattern has always been "create remote tempdir, run module_replacer to splat in common code, push generated common+module code to remote tempdir, execute module code on remote, capture stdout, delete tempdir". This is complicated further by things like async (which classically used a wrapper script to background the process and disconnect from the ssh/WinRM session so it could run after disconnection) and become (which hasn't been supported on Windows until now because it's a PITA). With pipelining, we just send the code to a remote interpreter in "get the code in stdin mode". With pipelining in place for WinRM, we skip a number of those steps and just exec a dumb wrapper stub that knows how to receive a JSON payload on stdin (which can optionally contain a chain of other wrappers to execute like become, async, persist, etc, as well as the module code and common module API code). From that JSON payload, it executes the wrapper chain, which does whatever special stuff the task config requires (disconnection, identity change, whatever), and at the bottom creates a PS runspace and gens up dynamic Powershell modules in-memory to execute the module code. We save several round-trips and disk I/O, making the module exec overhead much smaller.

**Comment by nitzmahone** [comment]
PS: this code is still suboptimal; lots of stuff that could be shared between wrappers, hacks, poor error handling/recovery, etc, but first priority was to make it work in time for 2.3. :)

**Comment by dagwieers** [comment]
I would prefer to see this hit devel as soon as possible so it receives wider testing.
Obviously only if you feel confident and it survives CI :-)

**Comment by nitzmahone** [comment]
@dagwieers My goal is to merge tonight, as soon as CI is green. :)

**Comment by jborean93** [comment]
@nitzmahone, looking good, I will try and do some testing when you have merged into devel, good luck.

**Comment by nitzmahone** [comment]
@jborean93 thanks! A few of known issues that aren't currently covered by tests:
- async + become together is broken
- binary module support is busted ATM (gonna comment out in CI and fix after merge)
- persistence is not enabled
- arbitrary module_utils stuff hasn't been committed yet (still sitting on my machine- some internal discussion on location/methods)

PS, thanks for the NTLM encryption stuff on pywinrm- I'll play with that soon and see where we're at!

**Comment by dagwieers** [comment]
W00t !

**Comment by tpaz** [comment]
+1!!

File: lib/ansible/constants.py [hunk 1]
```diff
@@ -265,7 +265,7 @@ def load_config_file():
 # Become
 BECOME_ERROR_STRINGS      = {'sudo': 'Sorry, try again.', 'su': 'Authentication failure', 'pbrun': '', 'pfexec': '', 'doas': 'Permission denied', 'dzdo': '', 'ksu': 'Password incorrect'} #FIXME: deal with i18n
 BECOME_MISSING_STRINGS    = {'sudo': 'sorry, a password is required to run sudo', 'su': '', 'pbrun': '', 'pfexec': '', 'doas': 'Authorization required', 'dzdo': '', 'ksu': 'No password given'} #FIXME: deal with i18n
-BECOME_METHODS            = ['sudo','su','pbrun','pfexec','doas','dzdo','ksu']
+BECOME_METHODS            = ['sudo','su','pbrun','pfexec','doas','dzdo','ksu','runas']
```
Comment (jamescassell) [inline lib/ansible/constants.py:268]: Every BECOME_METHOD except runas has a BECOME_ERROR_STRINGS and BECOME_MISSING_STRINGS.  Is it expected to exist for 'runas', too?

Comment (jborean93) [reply lib/ansible/constants.py:268]: Yes, runas is different from most of the other become methods. If you are having an issue with it, please raise a GH issue with the details.

**Pull request merged**